### PR TITLE
Add possibility to change the display mode (atom <-> structure)

### PR DESCRIPTION
**Pull request by sofiia-chorna**
Where to play around with modes:
```
npm i
npm run start
```
Open the `Chemical Shieldings` example

```
tox -e docs
python3 -m http.server
```
Open `examples/2-structure_map.html`

TODO:
- [x] change in EnvironmentInfo
- [x] change in map
- [x] change in structure
- [x] move mode out of indexer
- [x] add loader
- [x] change UI of toggle
- [x] testing (played with the examples from chemiscope.org & in doc/examples)

After review:
- [x] UI: add titles to buttons; substruct height of buttons in the map; reset color range
- [x] Code: document, delete usage of settings while changing the mode; fully recreate mapOptions
- [x] ~~Display errors as errors instread of warning~~ --> throw error, change logging in separate PR
- [x] Add `mode` to the settings

Note for exporting/importing of the dataset:
- If mode is changed and z-axes is not selected, we export NaN to JSON as `z.min` and `z.max`; hence it breaks once someone wants to import this dataset to the chemiscope. To do the import, it is necessary to delete the z-axes from the json settings (or set 0).
- This bug to be fixed in the separate PR

### CODE REVIEW

**Review by Luthaf — COMMENTED** [review]
This is working very well! The one small bug I found was that the color bar range is not reset when switching mode.

I've updated the example dataset for CSD to include a read structure PCA (the previous data only contained a single PCA dimension, with the other two being 0 everywhere); you can get the updated version by removing `app/example` and starting the dev server again.

File: src/info/info.ts [hunk 1]
```diff
@@ -57,8 +57,11 @@ export class EnvironmentInfo {
     private _shadow: ShadowRoot;
     private _root: HTMLElement;
     private _atom?: Info;
-    private _structure: Info;
+    private _structure!: Info;
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/info/info.ts:60]: I used ! it as `_structure` is initialised not directly in the constructor but in the function called in constructor -> TS complains that `_structure` might be undefined

File: src/info/info.ts [hunk 2]
```diff
@@ -387,12 +413,17 @@ export class EnvironmentInfo {
...
-            indexes = this._indexer.from_structure_atom(structure);
+            assert(!this._isParAtoms());
+            indexes = this._indexer.fromStructureAtom(this._mode, structure);
         }
         assert(indexes !== undefined);
         return indexes;
     }
+
+    /** Flag to determine if the current display mode is par atoms */
+    private _isParAtoms(): boolean {
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/info/info.ts:426]: Hehe good, I aggree

[the remark this answers is not in the document]
Comment (ceriottm) [reply src/info/info.ts:426]: random intervention - it should be _per_ atom, and I'd recommend trying to be as consistent as possible with the Structure/Environment nomenclature - so maybe the flag could be `envView` or something like that

File: src/map/map.ts [hunk 3]
```diff
@@ -216,29 +216,40 @@ export class PropertiesMap {
...
     /// Button used to reset the range of color axis
     private _colorReset: HTMLButtonElement;
     /// Plotly fix instance
     private _plotFix!: ReturnType<typeof fixPlot>;
+    /// Settings of config
+    private _configSettings: Settings;
+    /// Settings related to the stucture properties
+    private _structureSettings?: Settings | undefined;
+    /// Settings related to the atom properties
+    private _atomSettings?: Settings | undefined;
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/map/map.ts:232]: The idea was to calculate the atom/strucuture settings one time and then set `_options` to one of those variables.
Maybe I could simplify the logic...

File: src/map/options.ts [hunk 4]
```diff
@@ -96,69 +98,25 @@ export class MapOptions extends OptionsGroup {
...
-        root.appendChild(this._openModal);
+    /**
+     * Updates the map options with new properties and settings.
+     * Removes the old modal, reinitializes properties, and applies new settings.
+     * Instead of updating, the modal is re-created for efficient change
+     *
+     * @param {NumericProperties} properties new numeric properties
+     * @param {Settings} settings new settings
+     */
+    public updateMapOptions(properties: NumericProperties, settings: Settings) {
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/map/options.ts:114]: Hmm maybe I don't see other way how we can set new values to the settings (options in modals). My idea was to retrieve the current settings, update the necessary values, and set the state again.

However if settings are only needed for the initial load and not for subsequent updates, directly manipulating the class properties might work? Please let me know if you see a better way to handle them ^^

File: src/map/map.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply src/map/map.ts:232]: I don't understand. The settings are only applied once, when loading the dataset. Why do you need to keep them around?

[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/map/map.ts:232]: [update]
After a quick discussion, I won't touch the settings, so they will be reset by changing the mode:
atom (user settings applied) -> structure (no user settings) -> atom (no user settings)

File: src/info/info.ts [hunk 5]
```diff
@@ -57,8 +57,11 @@ export class EnvironmentInfo {
     private _shadow: ShadowRoot;
     private _root: HTMLElement;
     private _atom?: Info;
-    private _structure: Info;
+    private _structure!: Info;
     private _indexer: EnvironmentIndexer;
+    private _properties: { [name: string]: Property };
+    private _parameters?: { [name: string]: Parameter };
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/info/info.ts:63]: To re/initialise `_structure` and `_atom` once mode is changed outside of the constructor => recreate the part with the buttons (structure & atoms) in the info

File: src/map/options.ts [hunk 6]
```diff
@@ -64,30 +64,32 @@ export class AxisOptions extends OptionsGroup {
...
-    public z: AxisOptions;
-    public palette: HTMLOption<'string'>;
-    public symbol: HTMLOption<'string'>;
-    public color: {
+    public x!: AxisOptions;
+    public y!: AxisOptions;
+    public z!: AxisOptions;
+    public palette!: HTMLOption<'string'>;
+    public symbol!: HTMLOption<'string'>;
+    public color!: {
```
[the remark this answers is not in the document]
Comment (sofiia-chorna) [reply src/map/options.ts]: Done

**Comment by ceriottm** [comment]
had a quick look (will try to break it some more) - I know you've been talking about this already, but I noticed that it looks like settings for the structure viewers are preserved when you switch mode (e.g. if you set atoms to space filling, they remain space filling in both struc and env mode) while the map is reset. I find this very confusing.
Other quick comment: I think `Environment` is better than `Atoms` to indicate the local view mode.

**Comment by sofiia-chorna** [comment]
> had a quick look (will try to break it some more) - I know you've been talking about this already, but I noticed that it looks like settings for the structure viewers are preserved when you switch mode (e.g. if you set atoms to space filling, they remain space filling in both struc and env mode) while the map is reset. I find this very confusing. Other quick comment: I think `Environment` is better than `Atoms` to indicate the local view mode.

@ceriottm 
Thanks a lot for the feedback! Indeed, I could reset all settings of structure viewer too, if now it looks confusing, and rename Atoms to Environments

File: src/mode-toggle.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    /// Callback fired when the user changes the toggle value
+    public onchange: (mode: DisplayMode) => void;
+
+    /**
+     * Create a new {@link DisplayModeToggle} instance
+     *
+     * @param element HTML element or HTML id of the DOM element where the toggle will be attached
+     * @param isPerAtom flag indicating if the atom mode should be checked
+     */
+    constructor(element: string | HTMLElement, isPerAtom: boolean = true) {
```
Comment (Luthaf) [inline src/mode-toggle.ts:27]: instead of `isPerAtom`, why not pass the `mode` around here? (same in the function below)

**Review by Luthaf — COMMENTED** [review]
The code looks good! I like that you moved a bunch of functionalities to async, this will be nicer long term.

I'll try to break it a bit, and when @ceriottm is also done with trying to break the code we can merge this.

**Comment by ceriottm** [comment]
I tweaked the style a bit, and removed the title that seemed heavy and out of style with the rest of the UI.
I noticed a strange issue - if you fire up the website, then load the CS example and then switch back to say arginine, the space occupied by the toggler is not claimed back by the map, even though the div _is_ removed. 

Also, once we're done with the polish and the stress-testing, I think it'd be worth adding a sentence or two to the docs somewhere mentioning one can add both types of data to a json, and switch between the two.

**Comment by sofiia-chorna** [comment]
> I tweaked the style a bit, and removed the title that seemed heavy and out of style with the rest of the UI. I noticed a strange issue - if you fire up the website, then load the CS example and then switch back to say arginine, the space occupied by the toggler is not claimed back by the map, even though the div _is_ removed.
> 
> Also, once we're done with the polish and the stress-testing, I think it'd be worth adding a sentence or two to the docs somewhere mentioning one can add both types of data to a json, and switch between the two.

@ceriottm Thanks for checking! I pushed the fix for this issue you described with having the correct height back in map.

**Comment by ceriottm** [comment]
I've one small and one big problem:
* small problem: the fact that this is called "mode" switch is an issue because in python we use "mode" to refer to the type of viewer layout (map/structure/default). this is true in the source and in the settings (that btw still uses "atom" rather than "environment". I suggest we rename mode->target throughout
* bigger problem: I can't seem to be able to create (or interact with) two-mode chemiscopes from jupyter notebooks. it doesn't show the toggle (which is maybe OK, we don't want to clutter too much the jupyter viewers), it doesn't respond to changing the settings "mode" field (which is confusing), and when I export what I believe should be a two-mode chemiscope json, I can't open it.

**Comment by sofiia-chorna** [comment]
> * small problem: the fact that this is called "mode" switch is an issue because in python we use "mode" to refer to the type of viewer layout (map/structure/default). this is true in the source and in the settings (that btw still uses "atom" rather than "environment". I suggest we rename mode->target throughout

@ceriottm Thanks for taking a look!
- small issue:
I agree that renaming `mode` to `target` would help avoid confusion.
Regarding atom or environment, I am wondering about the consistency between this new `target` property (which refers to the display mode) and the target property used within `properties`, which can be either `atom` or `structure`.
If we adopt `settings.target` with values like `environment` or `structure`, should we align the values in `settings.properties` as well? Specifically, should `settings.properties` use `environment` instead of `atom` to maintain consistency?

- big issue: Thanks for checking it, let me take a look on the fix

**Comment by ceriottm** [comment]
> * small issue:
>   I agree that renaming `mode` to `target` would help avoid confusion.
>   Regarding atom or environment, I am wondering about the consistency between this new `target` property (which refers to the display mode) and the target property used within `properties`, which can be either `atom` or `structure`.
>   If we adopt `settings.target` with values like `environment` or `structure`, should we align the values in `settings.properties` as well? Specifically, should `settings.properties` use `environment` instead of `atom` to maintain consistency?

Ouch. I didn't realize we were using "atom". This is annoying because I'd say we don't want to lose backward compatibility over this. @Luthaf how would you see using and documenting target: "environment" but accepting also target: "atom"?

**Comment by Luthaf** [comment]
I'm not sure about `target: environment`. For most people, the `atom`/`structure` distinction is a lot clearer. The "environment" name was initially introduced for displaying **atom**-centered environments in the structure viewer. Unless there is a good reason to change it, I would keep the "atom" name for target everywhere. 

I agree on renaming "mode" to "target", since this matches the name we already use for the properties.

**Comment by ceriottm** [comment]
ok but then we should be consistent with environments -> atom everywhere.
TBD

**Comment by sofiia-chorna** [comment]
@ceriottm @Luthaf
I pushed a small fix to make sure the provided settings correspond to the settings' target.
Besides, here is how I re-install the chemiscope to have the version which is in development in the jupyter notebook:
```
pip wheel . --no-deps --no-build-isolation --check-build-dependencies --wheel-dir {envtmpdir}/dist
pip install {envtmpdir}/dist/chemiscope-0.7.1-py3-none-any.whl --force-reinstall
jupyter notebook (or restart kernel)
```
(Let me know if there is a better way, please)

After this re-install, I tested with the example `2-structure_map`, seems OK (the display is changing with `settings.target` property in jupyter).

Please let me know if we need to see the buttons in the jupyter widget.

**Comment by ceriottm** [comment]
> After this re-install, I tested with the example `2-structure_map`, seems OK (the display is changing with `settings.target` property in jupyter).
> 
> Please let me know if we need to see the buttons in the jupyter widget.

Do you mean it should change if one changes the `settings` traitlet, or the it'll display the right thing if one .shows chemiscope with the right target setting? The latter works, the traitlet seems not to

**Comment by sofiia-chorna** [comment]
> 
> Do you mean it should change if one changes the `settings` traitlet, or the it'll display the right thing if one .shows chemiscope with the right target setting? The latter works, the traitlet seems not to

@ceriottm 
Yes I tested by setting the target directly in the settings of `.show`. I might not fully understand how to change the target using the traitlet then.... Could you please provide an example of how you attempted to change it?

**Comment by ceriottm** [comment]
chemiscope.show returns an object (which you can show with display(cs))
If you set cs.settings to a valid setting dictionary, it should (and does) automatically update stuff like map axes, coloring, style...
If you do cs.settings={"target":"atom"} it does nothing

**Comment by sofiia-chorna** [comment]
> chemiscope.show returns an object (which you can show with display(cs)) If you set cs.settings to a valid setting dictionary, it should (and does) automatically update stuff like map axes, coloring, style... If you do cs.settings={"target":"atom"} it does nothing

Thanks for explaining!
It does not work for the `settings.target` but it seems for other settings properties neither. I tried the following:

```
cs = chemiscope.show(
    frames, properties, environments, settings={
   "map":{
      "x":{
         "property":"atom PCA[1]",
         "scale":"linear"
      },
   },
})

cs
==> plot with x.property = PCA[1] and scale = linear

cs.settings = {
   "map":{
      "x":{
         "property":"atom PCA[2]",
         "scale":"log"
      },
   },
}

cs
==> plot was not changed
```

Please let me know if I do it correctly and if it does work on your side

**Comment by ceriottm** [comment]
So, most likely the string was invalid, or the range giving nans (a PCA with a log scale is likely to give problems).
I think we should trigger errors in these cases, but this is probably for another PR. Let's finish this without the traitlet integration and maybe look at better traitlets support in a subsequent PR.

**Review by ceriottm — APPROVED** [review]
We already had a lot of back-and-forth, and I did my own round of polish. This is good to go IMO

**Comment by ceriottm** [comment]
Just to say I'll merge given that it looks like this works well, and that further tweaks of the logging and/or the traitlet interface will better wait for separate PRs. #1 is fixed!

**Pull request merged**